### ISSUE TITLE
Support new System.Text.Json source generator

### ISSUE THREAD

**Issue by tranb3r** [tmenier/Flurl#777]
Hi,

Is there any plan to support the new `System.Text.Json` source generator, in a future version of `Flurl`?
https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-source-generator/

I think it would require new extensions, similar to `PostJsonAsync` or `ReceiveJson`, but with an additionnal parameter (either `JsonTypeInfo` or `JsonSerializerContext`).
Or maybe it's already supported and I'm missing something?

Thanks

**Comment by tmenier** [comment]
Not familiar with this, I'll read up on it when I have some time. In a nutshell, what are you looking to accomplish?

**Comment by tranb3r** [comment]
Well, basically this new SG eliminates runtime reflection (amongst other improvements).
Which is critical for linker and native-aot stuff (aka the future of .NET).
The article explains the details of it, we can discuss the changes after you read it, nothing urgent.

**Comment by wbaldoumas** [comment]
+1 to this one. 

Implementing overloads of methods like `GetJsonAsync` and `PostJsonAsync` which take in `JsonTypeInfo` or `JsonSerializerContext` as parameters will really help folks wanting to leverage source generators for improved JSON serialization performance, as well as AoT (like @tranb3r mentions).

**Comment by NotAsea** [comment]
I second this. Basically you create overloads that accept `JsonTypeInfo<T>` as last parameter and just pass directly to `Serialize` `Deserialize` method of STG, no other complex stufff involve and suddenly Flurl become safe to trim again, i even suggest consume `Stream` directly in your response deserialize as STG has async overload for consume `BodyStream` which mean more async friendlier

**Comment by tranb3r** [comment]
@tmenier 
With the release of net9, this feature becomes even more important.
Is there a plan to implement these changes and publish a new version of Furl soon?
Thanks

**Comment by tranb3r** [comment]
There is actually another solution, which consists in implementing an `ISerializer` that takes a `JsonSerializerContext` as constructor parameter, and then uses the `JsonSerializerOptions` to create a `JsonTypeInfo` in `Serialize` or `Deserialize`. No need for additionnal parameters in `PostJsonAsync` or `ReceiveJson`.

Here is my code:
```
internal class JsonSourceGenSerializer : ISerializer
    {
        private readonly JsonSerializerOptions _jsonSerializerOptions;

        public JsonSourceGenSerializer(JsonSerializerContext jsonSerializerContext)
        {
            _jsonSerializerOptions = new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true,
                TypeInfoResolver = jsonSerializerContext
            };
        }

        public string Serialize(object obj)
        {
            var jsonTypeInfo = _jsonSerializerOptions.GetTypeInfo(obj.GetType());
            return JsonSerializer.Serialize(obj, jsonTypeInfo);
        }

        public T Deserialize<T>(string s) => string.IsNullOrWhiteSpace(s) ? default : JsonSerializer.Deserialize(s, (JsonTypeInfo<T>)_jsonSerializerOptions.GetTypeInfo(typeof(T)));

        public T Deserialize<T>(Stream stream) => stream.CanSeek && stream.Length == 0
            ? default
            : JsonSerializer.Deserialize(stream, (JsonTypeInfo<T>)_jsonSerializerOptions.GetTypeInfo(typeof(T)));
    }

...

// get a jsonSerializerContext (from source generator)
// create flurlClient and set JsonSerializer with jsonSerializerContext
flurlClient.WithSettings(settings => settings.JsonSerializer = new JsonSourceGenSerializer(jsonSerializerContext))
```

**Comment by tmenier** [comment]
@tranb3r That looks great, hopefully others can benefit. I'll look at incorporating this into the core at some point. (Sorry, haven't had a lot of free time lately.)

**Comment by tmenier** [comment]
@tranb3r Actually, looking at this more closely, if the goal is to eliminate runtime reflection, isn't `obj.GetType()` going to be self-defeating?

**Comment by tranb3r** [comment]
> @tranb3r Actually, looking at this more closely, if the goal is to eliminate runtime reflection, isn't `obj.GetType()` going to be self-defeating?

Actually, obj.GetType() is not using reflection.

**Comment by tmenier** [comment]
@tranb3r Sorry for the delay, busy time of year but taking a fresh look at this and it's clicking now. I think you've demonstrated that Flurl can leverage source generation today with its existing hooks. I'd take that a step further and suggest that you probably don't even need a custom `ISerializer` implementation - I believe the [default serializer](https://github.com/tmenier/Flurl/blob/dev/src/Flurl.Http/Configuration/DefaultJsonSerializer.cs) can facilitate this as-is. Consider:
- It accepts an optional `JsonSerializerOptions` instance in its constructor which, as you've demonstrated, can be assigned a `JsonSerializerContext`.
- It always passes that `JsonSerializerOptions` instance to `JsonSerializer` in its `Serialize` and `Deserialize` implementations. If a custom resolver is assigned to that options instance (like what you've done), it seems logical to conclude that `JsonSerializer` will use it, right?

**Comment by tranb3r** [comment]
@tmenier 

Yes and no.
- Yes because the `DefaultJsonSerializer` uses the `JsonSerializerOptions` instance (which can be assigned a `JsonSerializerContext`).
- No because the `System.Text.Json` method which uses `JsonSerializerOptions` is not aot (or trimming) compliant, contrary to the one using `JsonTypeInfo`.

This is important, because one reason to use the json source generator (not the only one but probably the most important one) is to eliminate reflection in order to allow native-aot support.

---

The `DefaultJsonSerializer` uses `JsonSerializer.Serialize(obj, _options)` which has `RequiresXXX` attributes:
```
[RequiresUnreferencedCode(SerializationUnreferencedCodeMessage)]
[RequiresDynamicCode(SerializationRequiresDynamicCodeMessage)]
public static string Serialize<TValue>(TValue value, JsonSerializerOptions? options = null)
{
    JsonTypeInfo<TValue> jsonTypeInfo = GetTypeInfo<TValue>(options);
    return WriteString(value, jsonTypeInfo);
}
```
The source-gen serializer uses `JsonSerializer.Serialize(obj, jsonTypeInfo)` which has no attribute:
```
public static string Serialize(object? value, JsonTypeInfo jsonTypeInfo)
{
    if (jsonTypeInfo is null)
    {
        ThrowHelper.ThrowArgumentNullException(nameof(jsonTypeInfo));
    }

    jsonTypeInfo.EnsureConfigured();
    return WriteStringAsObject(value, jsonTypeInfo);
}
```

**Comment by tranb3r** [comment]
@tmenier 
I agree, we can use `DefaultJsonSerializer`, since the `JsonSerializerOptions` constructor parameter can be assigned a `JsonSerializerContext`. The `JsonSerializer` will use it for serialization and deserialization.

There is although one small difference.
The `JsonSerializer.Serialize(obj, _options)` has `RequiresUnreferencedCode/RequiresDynamicCode` attributes, so there will be a compilation warning when enabling `IsAotCompatible`, since the compiler does not know whether options have been assigned a serializer context or not.
Of course this is not an issue if it's properly documented, but a `JsonSourceGenSerializer` makes it more explicit.

**Comment by tmenier** [comment]
Makes sense. This has been educational. :)

Let's say the default serializer worked both ways, and we had some "if" logic to decide which overloaded of `JsonSerializer.Serialize` to call. Would you get the compiler warning anyway, or would the trimmer somehow be smart enough to know that the `if (serializerContext == null)` (or whatever) path will never get hit? If the existence of that code path is enough to trigger the warning (I suspect it is), then I'm wondering if it's even possible to modify the default serializer in a way that accommodates both scenarios without triggering a warning. I probably need to get a better understanding of how that trimming process works in order to consider the right enhancement.

**Comment by tranb3r** [comment]
Right now Flurl is not aot-compatible, and some warnings appear when the user enables trim analysis warnings.
However json serialization totally works if the user makes sure the serializer context is set in the serialization options.

If you want to make Flurl aot-compatible, you have to stop using methods that are unsafe.
I think that with an "if" logic in the default serializer, you would still have unsafe code and the associated warning.

The question is now: do you want Flurl to be aot-compatible?
Json serialization is an issue (but with a workaround). However it's maybe not the only issue.
The next step would probably be to set `IsAotCompatible=true` in Flurl and look at the warnings it triggers.
